### PR TITLE
SVG #14 Converting Screenshot into text

### PR DESCRIPTION
**Pull request by ahmetutku**
This pull request introduces a new replay strategy that utilizes the "Segment Anything with Masking" (SAM) algorithm to segment screenshots before performing optical character recognition (OCR) to generate text. I have done some initial testing on this approach, but further testing will be needed to assess its effectiveness in various contexts. The replay strategy then generates the next input event using the processed text and previously recorded input events.

Next step: I plan to integrate the GPT-J model to predict the properties of the next InputEvent based on the processed text.

I am open to any feedback on this approach. Thanks!

### CODE REVIEW

**Comment by abrichr** [comment]
Thanks @ahmetutku !

Can you please make your changes to a new strategy in a new file? Also, can you please clarify why you are using `pytesseract` and not re-using the existing `OCRReplayStrategyMixin`?

Your PR title ("SVG #14 Converting Screenshot into text") suggests you are interested in addressing https://github.com/MLDSAI/puterbot/issues/14, but your PR description ("Segment Anything with Masking") suggests you are interested in addressing https://github.com/MLDSAI/puterbot/issues/15. Can you please clarify?

> I am open to any feedback on this approach. Thanks!

I'm not sure there is enough here to provide feedback. Can you please clarify what you are trying to accomplish, and how you are thinking about go about it?

Thank you! 🙏

**Comment by ahmetutku** [comment]
Hello! 
I chose to use pytesseract because I found it to be faster and more accurate for my specific use case. I wanted to extract text from images in real-time and pytesseract offered faster processing speeds than the existing OCRReplayStrategyMixin. Additionally, pytesseract has a higher accuracy rate than the OCRReplayStrategyMixin in some cases, which was important for my project. However, I understand that OCRReplayStrategyMixin may be a better choice in certain situations, depending on the specific needs of the project.

I appologize for the cnfusion I am solely focusing on #14. I have  a new commit with svg.py in a seprate file that inculdes the following code: "import pytesseract
from PIL import Image

import torch
from transformers import AutoTokenizer, AutoModelForCausalLM

# Load pre-trained GPT-2 model
model_name = 'gpt2'
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

class OCRReplayMixin:
    def get_text_from_screenshot(self, screenshot):
        # Convert the screenshot to text using OCR
        text = pytesseract.image_to_string(screenshot)
        
        # Encode the OCR text using the tokenizer
        input_ids = tokenizer.encode(text, return_tensors='pt')
        
        # Generate text using the pre-trained model
        output_ids = model.generate(input_ids=input_ids, max_length=100)
        output_text = tokenizer.decode(output_ids[0], skip_special_tokens=True)
        
        return output_text"
I had to use GTP 2 because GTP 3 was not accepting any new users & the one you have recommended GTP4ALL required a higher version of macOS than what I had. I have another question now. If I am not supposed to modify the given files. How would I integrate my code into the program to check if it works? Please note that the code is not finalized and may contain some errors. Thank you!

**Comment by abrichr** [comment]
Hi @ahmetutku , thank you for clarifying!

> faster and more accurate

If you get a chance can you please create a new issue comparing the speed and accuracy?

(I edited your comment so that it's easier to read. For output like you have above, it's best to surround it with triple-backticks before and after; see https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#fenced-code-blocks)

The code above does not appear to involve SVG. To clarify, SVG stands for Scalable Vector Graphic. Consider https://pypi.org/project/pypotrace/ (or similar) for creating SVG.

> I had to use GTP 2 because GTP 3 was not accepting any new users

The choice of model is up to you. However can you please clarify exactly what you mean by "GTP 3 was not accepting any new users"?

> If I am not supposed to modify the given files.

Please feel free to modify any files you wish!

**Comment by abrichr** [comment]
Hi @ahmetutku , almost there! Please follow the instructions at https://github.com/MLDSAI/puterbot/pull/55#issuecomment-1530757878  but modified for your fork, e.g.:

```
git remote add ahmetutku https://github.com/ahmetutku/puterbot.git
git fetch ahmetutku
# what goes here to run your code? please create a branch update your PR to merge from that branch, e.g.:
git checkout feat/<your-feature-name>  # please fill this in
pip install -r requirements.txt  # please fill this in
python -m ...  # please fill this in
```

**Comment by abrichr** [comment]
@ahmetutku thank you for the great work! Please grab some time on my calendar at your earliest convenience on Weds/Thurs this week: https://www.getclockwise.com/c/richard-abrich/quick-meeting

**Comment by ahmetutku** [comment]
I apologize for the delay in my response. I wanted to provide you with more information regarding your questions:

1) "If you get a chance can you please create a new issue comparing the speed and accuracy?" I came across OCR and was introduced to OCR and Pytesseract through this link: "https://nanonets.com/blog/ocr-with-tesseract/". After doing some research, I found that Pytesseract is known for its speed, efficiency, and accuracy, and is often cited as one of the fastest and most accurate OCR libraries available. It is also known to be beginner-friendly, which is why I decided to use it.

2) "The code above does not appear to involve SVG. To clarify, SVG stands for Scalable Vector Graphic. Consider https://pypi.org/project/pypotrace/ (or similar) for creating SVG." Thank you for clarifying what SVG stands for. I assume we should convert the screenshot to SVG format. Is non-SVG format not supported in the OCRReplayMixin we have? Additionally, I am curious if we are using SVG format for scalability and memory space?

3) "The choice of model is up to you. However can you please clarify exactly what you mean by "GTP 3 was not accepting any new users"?" In my research, I discovered that to access GTP 3, I need an API key, which requires me to sign in. However, when I attempted to get an API key, the website indicated that it is not currently accepting new users. Fortunately, I have updated my computer, so I should be able to use the recommended link now.

Thank you for your help and guidance.

**Comment by abrichr** [comment]
> After doing some research, I found that Pytesseract is known for its
speed, efficiency, and accuracy, and is often cited as one of the fastest
and most accurate OCR libraries available. It is also known to be
beginner-friendly, which is why I decided to use it.

Can you please refactor this into ocr_mixin.py as a separate method?

> I assume we should convert the screenshot to SVG format.

Correct.

> Is non-SVG format not supported in the OCRReplayMixin we have

Can you please clarify your question? OCR and SVG are separate.

 > Additionally, I am curious if we are using SVG format for scalability
and memory space?

We need to convert the image to a format the model can understand. SVG is
one approach. They each have various tradeoffs and are likely suitable in
different contexts.

> In my research, I discovered that to access GTP 3, I need an API key,
which requires me to sign in. However, when I attempted to get an API key,
the website indicated that it is not currently accepting new users.
Fortunately, I have updated my computer, so I should be able to use the
recommended link now.

It's still not clear what you are referring to. "GPT" stands for Generative
Pretrained Transformers. Are you referring to OpenAI's API? You should not
need to access this, but if you want to implement it there is an open issue
here: https://github.com/MLDSAI/PuterBot/issues/63 . See
https://github.com/MLDSAI/PuterBot/blob/main/puterbot/strategies/llm_mixin.py
for an existing alternative.

**Comment by ahmetutku** [comment]
> Can you please refactor this into ocr_mixin.py as a separate method?
When you say refactor do you mean modify the ocr_mixin to include Pytesseract instead of RapidOCR? 

> Can you please clarify your question? OCR and SVG are separate.
> We need to convert the image to a format the model can understand. SVG is one approach. They each have various tradeoffs and are likely suitable in different contexts.

I understand now that we convert the screenshots & input events to SVG format so OCR algorithms accurately recognize and extract the text from the image. But this leaves me with the question: where are the actions recorded & converted into OCR? I am asking this question because I believe that we should convert screenshots & input e vents into SVG format before the OCR mixin is called. Is that correct? 

> It's still not clear what you are referring to. "GPT" stands for Generative Pretrained Transformers. Are you referring to OpenAI's API? You should not need to access this, but if you want to implement it there is an open issue here: #63 . See https://github.com/MLDSAI/PuterBot/blob/main/puterbot/strategies/llm_mixin.py for an existing alternative.

I apologize for the confusion; I have updated my computer, and I am able to use the recommended LLM now. Where would I feed the generated text into the LLM; more specifically, I am aware that I need to create a separate Python file that does the SVG conversion LLM feeding, but how will this file interact with the rest of the program? In which file would it be used?

**Comment by abrichr** [comment]
> where are the actions recorded

puterbot/record.py

> & converted into OCR?

puterbot/strategies/ocr_mixin.py

> I am asking this question because I believe that we should convert screenshots & input e vents into SVG format before the OCR mixin is called. Is that correct?

The order doesn't matter. What matter is what you do with the resulting SVG and OCR.

Please refer to the instructions in the README: https://github.com/mldsai/puterbot#contributing
